relays/substrate/src/messages_source.rs

```diff
@@ -0,0 +1,221 @@
+	message_lane::{MessageLane, SourceHeaderIdOf, TargetHeaderIdOf},
+	message_lane_loop::{ClientState, SourceClient, SourceClientState},
+};
+use relay_substrate_client::{Chain, Client, Error as SubstrateError, HashOf, HeaderIdOf};
+use relay_utils::HeaderId;
+use sp_core::Bytes;
+use sp_runtime::{traits::Header as HeaderT, DeserializeOwned};
+use sp_trie::StorageProof;
+use std::{marker::PhantomData, ops::RangeInclusive};
+
+/// Substrate client as Substrate messages source.
+pub struct SubstrateMessagesSource<C: Chain, P, M> {
+	client: Client<C>,
+	tx_maker: M,
+	lane: LaneId,
+	instance: InstanceId,
+	_marker: PhantomData<P>,
+}
+
+/// Substrate transactions maker.
+#[async_trait]
+pub trait SubstrateTransactionMaker<C: Chain, P: MessageLane>: Clone + Send + Sync {
+	/// Signed transaction type.
+	type SignedTransaction: Send + Sync + Encode;
+
+	/// Make messages receiving proof transaction.
+	async fn make_messages_receiving_proof_transaction(
+		&self,
+		generated_at_block: TargetHeaderIdOf<P>,
+		proof: P::MessagesReceivingProof,
+	) -> Result<Self::SignedTransaction, SubstrateError>;
+}
+
+impl<C: Chain, P, M> SubstrateMessagesSource<C, P, M> {
+	/// Create new Substrate headers source.
+	pub fn new(client: Client<C>, tx_maker: M, lane: LaneId, instance: InstanceId) -> Self {
+		SubstrateMessagesSource {
+			client,
+			tx_maker,
+			lane,
+			instance,
+			_marker: Default::default(),
+		}
+	}
+}
+
+impl<C: Chain, P, M: Clone> Clone for SubstrateMessagesSource<C, P, M> {
+	fn clone(&self) -> Self {
+		Self {
+			client: self.client.clone(),
+			tx_maker: self.tx_maker.clone(),
+			lane: self.lane,
+			instance: self.instance,
+			_marker: Default::default(),
+		}
+	}
+}
+
+#[async_trait]
+impl<C, P, M> SourceClient<P> for SubstrateMessagesSource<C, P, M>
+where
+	C: Chain,
+	C::Header: DeserializeOwned,
+	C::Index: DeserializeOwned,
+	<C::Header as HeaderT>::Number: Into<u64>,
+	P: MessageLane<
+		MessageNonce = MessageNonce,
+		MessagesProof = (Weight, (HashOf<C>, StorageProof, LaneId, MessageNonce, MessageNonce)),
+		SourceHeaderNumber = <C::Header as HeaderT>::Number,
+		SourceHeaderHash = <C::Header as HeaderT>::Hash,
+	>,
+	P::TargetHeaderNumber: Decode,
+	P::TargetHeaderHash: Decode,
+	M: SubstrateTransactionMaker<C, P>,
+{
+	type Error = SubstrateError;
+
+	async fn reconnect(mut self) -> Result<Self, Self::Error> {
+		let new_client = self.client.clone().reconnect().await?;
+		self.client = new_client;
+		Ok(self)
+	}
+
+	async fn state(&self) -> Result<SourceClientState<P>, Self::Error> {
+		read_client_state::<_, P::TargetHeaderHash, P::TargetHeaderNumber>(&self.client, P::TARGET_NAME).await
+	}
+
+	async fn latest_generated_nonce(
+		&self,
+		id: SourceHeaderIdOf<P>,
+	) -> Result<(SourceHeaderIdOf<P>, P::MessageNonce), Self::Error> {
+		let encoded_response = self
+			.client
+			.state_call(
+				// TODO: https://github.com/paritytech/parity-bridges-common/issues/457
+				"OutboundLaneApi_latest_generated_nonce".into(),
```

Comment:
We should do something about these hardcoded Runtime API strings, we already saw how easy they were to break last week

Reply:
IMHO it's worth to at least create constants and maybe even group in a single module, so that there is one single place we need to update in the future. This also encourages re-use if we ever run into using the same Runtime API call in different places

Reply:
This hasn't helped last time we had a name change. Probably better would be to introduce replacement for `decl_runtime_apis! {}` macro (that would actually call `decl_runtime_apis` from within)  that will generate string constants from method names, or even auto-generate `jsonrpsee` client for these methods. Re this particular string(s) - they're going to change anyway as a part of #457 in near future - I'll extract constants there.